src/tacacs_plus_fmt_plug.c

```diff
@@ -0,0 +1,337 @@
+ *    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ *    A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ *    OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ *    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ *    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ *    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_tacacsplus;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_tacacsplus);
+#else
+
+#include <string.h>
+#ifdef _OPENMP
+#include <omp.h>
+#ifndef OMP_SCALE
+#define OMP_SCALE               2048
+#endif
+#endif
+
+#include "formats.h"
+#include "misc.h"
+#include "common.h"
+#include "params.h"
+#include "options.h"
+#include "md5.h"
+#include "johnswap.h"
+#include "memdbg.h"
+
+#define FORMAT_LABEL            "tacacs-plus"
+#define FORMAT_NAME             "TACACS+"
+#define FORMAT_TAG              "$tacacs-plus$"
+#define TAG_LENGTH              (sizeof(FORMAT_TAG) - 1)
+#define ALGORITHM_NAME          "MD5 32/" ARCH_BITS_STR
+#define BENCHMARK_COMMENT       ""
+#define BENCHMARK_LENGTH        0
+#define PLAINTEXT_LENGTH        125
+#define BINARY_SIZE             0
+#define BINARY_ALIGN            1
+#define SALT_SIZE               sizeof(struct custom_salt)
+#define SALT_ALIGN              sizeof(uint32_t)
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      1
+
+#define MAX_CIPHERTEXT_LENGTH   1500
+#ifndef MD5_DIGEST_LENGTH
+#define MD5_DIGEST_LENGTH       16
+#endif
+
+static struct fmt_tests tests[] = {
+	{"$tacacs-plus$0$6d0e1631$db7c01e77499$c006", "1234"},
+	{"$tacacs-plus$0$6d0e1631$d623c7692ca7b12f7ecef113bea72845$c004", "1234"},
+	{"$tacacs-plus$0$6d0e1631$f7711e4b904fc4a4753e923e9bf3d2cc33e9febd3d2db74b9aa6d20462c2072013c77345d7112400d7b915$c002", "1234"},
+	{"$tacacs-plus$0$d0cb2225$f5acae34eb560981b30315cc9b08f00a31566eb5f09351713d0b33067ea1fbdb63ed84062ec24ebc45de63$c002", "testing123"},
+	{"$tacacs-plus$0$acf4c30b$c73c409532a4a80e58ba94391111e300$c002", "12345"},
+	{NULL}
+};
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static int any_cracked, *cracked;
+static size_t cracked_size;
+
+static struct custom_salt {
+	uint32_t ctlen;
+	MD5_CTX pctx;
+	uint32_t pre_hash_data_len;
+	unsigned char pre_hash_data[16];
+	uint32_t hash_data_len;
+	unsigned char hash_data[16];
+	unsigned char ciphertext[MAX_CIPHERTEXT_LENGTH];
+} *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+#ifdef _OPENMP
+	int omp_t = omp_get_max_threads();
+
+	self->params.min_keys_per_crypt *= omp_t;
+	omp_t *= OMP_SCALE;
+	self->params.max_keys_per_crypt *= omp_t;
+#endif
```

Comment:
Please switch this to the new approach I proposed in the john-dev posting above, where the multiplicatons are only reached when we're actually running more than 1 thread. Indeed, this change goes along with the correction to `OMP_SCALE` vs. `MAX_KEYS_PER_CRYPT` split mentioned above, so you must do both at once.